src/Microsoft.AspNetCore.Razor.Runtime/Runtime/TagHelpers/TagHelperDescriptorFactory.cs

```diff
@@ -750,5 +733,311 @@ private static string ToHtmlCase(string name)
         {
             return HtmlCaseRegex.Replace(name, HtmlCaseRegexReplacement).ToLowerInvariant();
         }
+
+        // Internal for testing
+        internal class RequiredAttributeParser
+        {
+            private static readonly char[] InvalidPlainAttributeNameCharacters = { ' ', '\t', ',', RequiredAttributeWildcardSuffix };
+            private static readonly char[] InvalidCSSAttributeNameCharacters = (new[] { ' ', '\t', ',', ']' }).Concat(TagHelperRequiredAttributeDescriptor.SupportedCSSValueOperators).ToArray();
+            private static readonly char[] InvalidCSSQuotelessValueCharacters = { ' ', '\t', ']' };
```

Comment:
Of course, but that's why I have `PlainAttributeNameCharacters` etc. as the suffixes to say they're invalid for a specific portion of the required attribute.